doc/xboxdrv.1
.\" t
.\" -*- coding: us-ascii -*-
.if \n(.g .ds T< \\FC
.if \n(.g .ds T> \\F[\n[.fam]]
.de URL
\\$2 \(la\\$1\(ra\\$3
..
.if \n(.g .mso www.tmac
.TH "xboxdrv " 1 2011-07-20 0.8.9 "User Commands"
.SH NAME
xboxdrv
\- A Xbox/Xbox360 gamepad driver that works in userspace 
.SH SYNOPSIS
'nh
.fi
.ad l
\fBxboxdrv\fR \kx
.if (\nx>(\n(.l/2)) .nr x (\n(.l/5)
'in \n(.iu+\nxu
[OPTION]\&... [--] [COMMAND] [ARGUMENTS]
'in \n(.iu-\nxu
.ad b
'hy
.SH DESCRIPTION
\fBxboxdrv\fR is a driver for Xbox and Xbox360
gamepads. It works by reading the raw data from the controller
with the userspace library libusb and then passes the
interpreted data to the kernel via uinput. This
allows \fBxboxdrv\fR to provide regular joystick
and event devices, which makes it compatible with all Linux
software.
.PP
Aside from the pure driver, \fBxboxdrv\fR also
includes a rich set of configuration options that allow you to
tweak the abilities of the virtual input devices that xboxdrv
will create. This includes basic button and axis remapping, as
well as more complicated things like mouse and keyboard emulation,
auto-fire and throttle control emulation.
.PP
It is also possible for \fBxboxdrv\fR to read input
data directly from an event device, this allows the use of the
configurability of \fBxboxdrv\fR on regular PC
joysticks, keyboards and mice and thus
lets \fBxboxdrv\fR serve a similar purpose
as \fBjoy2key\fR. See the
option \*(T<\fB\-\-evdev\fR\*(T> below for more information.
.PP
When a \*(T<\fBCOMMAND\fR\*(T> is provided xboxdrv will launch
that application and be running till that application exits.
This is a convenience function to make it easier to use xboxdrv
in wrapper scripts. See the section
[Writing Start-Up Scripts for Games]
for more information.
.SH OPTIONS
.SS "GENERAL OPTIONS"
.TP 
\*(T<\fB\-h\fR\*(T>, \*(T<\fB\-\-help\fR\*(T>
Display help text and exit.
.TP 
-V, --version
Print the version number and exit.
.TP 
\*(T<\fB\-v\fR\*(T>, \*(T<\fB\-\-verbose\fR\*(T>
Print verbose messages.
.TP 
\*(T<\fB\-\-debug\fR\*(T>
Print even more verbose messages then \*(T<\fB\-\-verbose\fR\*(T>.
.TP 
\*(T<\fB\-s\fR\*(T>, \*(T<\fB\-\-silent\fR\*(T>
Do not display controller events on the terminal. For
regular use this option should always be used as output
data to the terminal can eat up quite a bit of CPU.
.TP 
\*(T<\fB\-\-quiet\fR\*(T>
Do not display startup text and suppress most other output.
.TP 
\*(T<\fB\-\-priority\fR\*(T> \fIPRIORITY\fR
Possible values for \fIPRIORITY\fR
are "normal" and "realtime". Realtime scheduling gives
the xboxdrv process higher priority and thus allows it
to function properly even when the machine is under
load.

Note that realtime priority requires running xboxdrv as
root, when running xboxdrv as user there is no way to
increase the priority.

This option is deprecated,
use \fBchrt\fR(1)
instead to achive the same effect.
.SS "LIST OPTIONS"
.TP 
\*(T<\fB\-\-help\-led\fR\*(T>
List possible values for the led.
.TP 
\*(T<\fB\-\-help\-devices\fR\*(T>
List supported devices.
.TP 
\*(T<\fB\-\-list\-supported\-devices\fR\*(T>
List supported devices (used by xboxdrv-daemon.py).
.TP 
\*(T<\fB\-\-list\-supported\-devices\-xpad\fR\*(T>
List supported devices in \*(T<\fIxpad.c\fR\*(T> style.
.TP 
\*(T<\fB\-\-help\-abs\fR\*(T>
List all allowed EV_ABS symbols.
.TP 
\*(T<\fB\-\-help\-rel\fR\*(T>
List all allowed EV_REL symbols.
.TP 
\*(T<\fB\-\-help\-key\fR\*(T>
List all allowed EV_KEY symbols.
.TP 
\*(T<\fB\-\-help\-x11keysym\fR\*(T>
List all allowed X11 Keysym symbols.
.TP 
\*(T<\fB\-\-help\-axis\fR\*(T>
List all allowed XboxAxis symbols.
.TP 
\*(T<\fB\-\-help\-button\fR\*(T>
List all allowed XboxButton symbols.
.TP 
\*(T<\fB\-\-help\-all\fR\*(T>
List all symbols that can be used
in \*(T<\fB\-\-ui\-buttonmap\fR\*(T>, \*(T<\fB\-\-ui\-axismap\fR\*(T>, \*(T<\fB\-\-buttonmap\fR\*(T>
and \*(T<\fB\-\-axismap\fR\*(T>. This option is the same as
\*(T<\fB\-\-help\-abs\fR\*(T>,
\*(T<\fB\-\-help\-rel\fR\*(T>,
\*(T<\fB\-\-help\-key\fR\*(T>,
\*(T<\fB\-\-help\-x11keysym\fR\*(T>,
\*(T<\fB\-\-help\-axis\fR\*(T> and
\*(T<\fB\-\-help\-button\fR\*(T>.
.SS "CONFIG FILE OPTIONS"
.TP 
\*(T<\fB\-c\fR\*(T>, \*(T<\fB\-\-config\fR\*(T> \fIFILE\fR
Reads configuration information from \fIFILE\fR.
Configurations from file are handling as if they would
be command line options at the position
of \*(T<\fB\-\-config\fR\*(T> \fIFILE\fR.

The syntax
of \fIFILE\fR is
the familiar INI syntax used for many configuration
files. Regular key/value pairs must go into the
[xboxdrv] section. '#' and ';' can be used for comments.
Key names have for most part the same name as command
line options. Command line options that take a list of
input mappings (--ui-buttonmap, --ui-axismap,
--evdev-absmap, ...) can be split of into their own
section for better readability.

The \*(T<\fIexamples/\fR\*(T> directory contains
some example configuration files.

.nf
\*(T<[xboxdrv]
silent=true
deadzone=6000
dpad\-as\-button=true
trigger\-as\-button=true

[ui\-axismap]
x2=REL_X:10
y2=REL_Y:\-10
x1=KEY_A:KEY_D
y1=KEY_W:KEY_S

[ui\-buttonmap]
a=KEY_LEFTSHIFT
b=BTN_C
x=BTN_EXTRA
y=KEY_C

[ui\-buttonmap]
lb=BTN_RIGHT
rb=KEY_SPACE

[ui\-buttonmap]
lt=KEY_Z
rt=BTN_LEFT

[ui\-buttonmap]
dl=KEY_4
dr=KEY_2
du=REL_WHEEL:\-1:150
dd=REL_WHEEL:1:150

[ui\-buttonmap]
back=KEY_TAB
start=KEY_ESC

# EOF #\*(T>
.fi
.TP 
\*(T<\fB\-\-alt\-config\fR\*(T> \fIFILE\fR
A shortcut for
writing \*(T<\fB\-\-next\-config\fR\*(T> \*(T<\fB\-\-config\fR\*(T> \fIFILE\fR.

To load multiple configuration options use:

.nf
\*(T<xboxdrv \-\-config first.ini \-\-alt\-config second.ini \-\-alt\-config third.ini\*(T>
.fi
.TP 
\*(T<\fB\-o\fR\*(T>, \*(T<\fB\-\-option\fR\*(T> \fINAME=VALUE\fR
Set an option as if it would come from a config file from the command line.
.TP 
\*(T<\fB\-\-write\-config\fR\*(T> \fIFILE\fR
Write an example configuration file to \fIFILE\fR.
.SS "DAEMON OPTIONS"
.TP 
\*(T<\fB\-D\fR\*(T>, \*(T<\fB\-\-daemon\fR\*(T>
Run xboxdrv as daemon. If this option is given xboxdrv
will listen to udev for USB connection events and launch
driver threads for newly connected controllers.

Configuration options can still be supplied as usual.
Note however that xboxdrv when run as daemon will not
create new uinput devices on demand, instead it will
only create devices once at startup for the given
configurations and then assign new controllers to these
configurations. While this means xboxdrv can't support
an unlimited number of controllers, it also means that
xboxdrv can allow hot plugging even for applications
that don't support it themselves, as applications will
only see the permanent device files, not the controller
that xboxdrv will change around under the hood.

An example configuration that supports three controller
would look like this:

.nf
\*(T<xboxdrv \-\-daemon \e
    # config options for the first controller
  \-\-next\-controller \e
    # config options for the second controller
  \-\-next\-controller
    # config options for the third controller\*(T>
.fi

The \*(T<\fB\-\-match\fR\*(T> option can be used to limit
the controller slots to only those controllers that
match the given RULE and thus be used to assign
configurations only to specific controllers.
.TP 
\*(T<\fB\-\-detach\fR\*(T>
Detaches xboxdrv from the current shell, only valid
if \*(T<\fB\-\-daemon\fR\*(T> is given.
.TP 
\*(T<\fB\-\-pid\-file\fR\*(T> \fIFILE\fR
Write the xboxdrv daemon process id to FILE.
.TP 
\*(T<\fB\-\-dbus\fR\*(T> \fIBUS\fR
Set which bus type xboxdrv should connect to. Allowed
values for BUS are session,
system,
disabled and
auto. The default is
auto, which will detect the
appropriate bus type depending on if xboxdrv is run as
root (system or as user
(session). Running with
disabled will disable D-Bus support
completely.
.TP 
\*(T<\fB\-\-on\-connect\fR\*(T> \fIEXE\fR
Launches \fIEXE\fR
when a controller gets connected. As arguments
"\fIBUSDEV\fR:\fIDEVNUM\fR",
"\fIidVendor\fR:\fIidProduct\fR",
"\fINAME\fR are provided.
.TP 
\*(T<\fB\-\-on\-disconnect\fR\*(T> \fIEXE\fR
Launches \fIEXE\fR
when a controller gets disconnected. As arguments
"\fIBUSDEV\fR:\fIDEVNUM\fR",
"\fIidVendor\fR:\fIidProduct\fR",
"\fINAME\fR are provided.
.SS "DEVICE OPTIONS"
.TP 
\*(T<\fB\-L\fR\*(T>, \*(T<\fB\-\-list\-controller\fR\*(T>
List available controllers on the system.
.TP 
\*(T<\fB\-i\fR\*(T>, \*(T<\fB\-\-id\fR\*(T> \fIN\fR
Use controller with id N (default: 0),
use \*(T<\fB\-\-list\-controller\fR\*(T> to obtain a list
of available controller.
.TP 
\*(T<\fB\-w\fR\*(T>, \*(T<\fB\-\-wid\fR\*(T> \fIN\fR
Use wireless controller with wid N (default: 0).
.TP 
\*(T<\fB\-\-device\-by\-path\fR\*(T> \fIBUS:DEV\fR
Use the controller at BUS:DEV, do not do any automatic
scanning. Useful for cases when a controller isn't known
by xboxdrv, but supports one of the given protocols.
.TP 
\*(T<\fB\-\-device\-by\-id\fR\*(T> \fIVENDOR:PRODUCT\fR
Use device that matches VENDOR:PRODUCT (as returned
by \fBlsusb\fR). Useful for cases when a
controller isn't known by xboxdrv, but supports one of
the given protocols.
.TP 
\*(T<\fB\-\-type\fR\*(T> \fITYPE\fR
Ignore autodetection and enforce the controller type.
Possible values
for \fITYPE\fR:
.RS 
.TP 0.2i
\(bu
xbox
.TP 0.2i
\(bu
xbox-mat
.TP 0.2i
\(bu
xbox360
.TP 0.2i
\(bu
xbox360-wireless
.TP 0.2i
\(bu
xbox360-guitar
.TP 0.2i
\(bu
firestorm
.TP 0.2i
\(bu
firestorm-vsb
.TP 0.2i
\(bu
t-wireless
.TP 0.2i
\(bu
saitek-p2500
.TP 0.2i
\(bu
generic-usb
.RE

The \*(T<\fBgeneric\-usb\fR\*(T> type is a special type
that will work with any USB controller, it will however
not interpret the input it gets, but just dump it to the
console for development purposes.
See \*(T<\fB\-\-generic\-usb\-spec\fR\*(T> for further
information.
.TP 
\*(T<\fB\-d\fR\*(T>, \*(T<\fB\-\-detach\-kernel\-driver\fR\*(T>
Detaches the kernel driver that is currently associated
with the given device. This is useful when you have the
xpad module loaded and want to use xboxdrv without
unloading it.
.TP 
\*(T<\fB\-\-generic\-usb\-spec\fR\*(T> \fINAME=VALUE,...\fR
Allows to specify from which
endpoint \*(T<\fBgeneric\-usb\fR\*(T> will read. The
spec as the form
of \fINAME=VALUE,...\fR. Allowed values are:
.RS 
.TP 
vid=HEX
The vendor id of the controller to which this spec applies
.TP 
pid=HEX
The product id of the controller to which this spec applies
.TP 
if=NUM
The interface from which GenericUSBController should be read
.TP 
ep=NUM
The endpoint from which GenericUSBController should be read
.RE
.SS "EVDEV OPTION"
.TP 
\*(T<\fB\-\-evdev\fR\*(T> \fIDEVICE\fR
Allows you to read input data from a regular event
device. This allows you to
use \fBxboxdrv\fR on regular PC
joysticks. The data that is read from the event device
is converted internally into a XboxMsg object and then
passed through the same configuration pipeline as it
would be for a regular Xbox360 controller. This allows
you to make use of all the regular configurability, but
limits you to the number of axis and buttons that an
Xbox360 controller provides.

As a regular PC joystick will most likely already create
a \*(T<\fI/dev/input/jsX\fR\*(T> device by itself,
you might need to get rid of that so that a game will
properly detect the joystick device created
by \fBxboxdrv\fR. The easiest way to
accomplish that is to simply delete the old joystick and
rename the device that \fBxboxdrv\fR
created to \*(T<\fI/dev/input/js0\fR\*(T>. When you
use udev, this operation should be harmless and
automatically reverse itself when you remove the
controller and plug it back in or when you reboot the
computer.
.TP 
\*(T<\fB\-\-evdev\-debug\fR\*(T>
The evdev event handler will print all received events
to stdout, this makes it easy to see which events a
given controller sends.
.TP 
\*(T<\fB\-\-evdev\-no\-grab\fR\*(T>
By default the evdev driver will grab the device, thus
making it impossible for other applications to receive
events from that device. This is done to avoid confusing
applications, as otherwise an app would receive every
event twice, once from the original device and once from
the virtual xboxdrv one. In some cases this behaviour is
undesired, such when mapping only an otherwise
unhandled subset of keys of an device, i.e. mapping the
multimedia keys on a keyboard, so this option turns the
grab off.
.TP 
\*(T<\fB\-\-evdev\-absmap\fR\*(T> \fIABSMAP,...\fR
.nf
\*(T<ABSMAP = EVDEV_ABS [ "+", "\-" ] "=" XBOXAXIS ;\*(T>
.fi

Sets how evdev events are mapped to Xbox axis
events. An example configuration would look like this:

.nf
\*(T<\-\-evdev\-absmap ABS_X=x1,ABS_Y=y1,ABS_RZ=x2,ABS_THROTTLE=y2,ABS_HAT0X=dpad_x,ABS_HAT0Y=dpad_y\*(T>
.fi

\fBxboxdrv\fR will output on startup a full
list of event names that the given event device
supports and that can be used in place of \fIEVDEV_ABS\fR.

It is also possible to map half-axis with a command like:

.nf
\*(T<\-\-evdev\-absmap ABS_Y+=LT,ABS_Y\-=RT\*(T>
.fi

This will map the upward movement of the Y axis to the
left trigger and the downward movement to the right
trigger. 
.TP 
\*(T<\fB\-\-evdev\-keymap\fR\*(T> \fIKEYMAP\fR
Sets how evdev events are mapped to Xbox controller
events. An example configuration would look like this:

.nf
\*(T<\-\-evdev\-keymap BTN_TRIGGER=a,BTN_THUMB=b,BTN_THUMB2=x\*(T>
.fi

\fBxboxdrv\fR will output on start a full
list of event names that the given event device
supports.
.SS "STATUS OPTIONS"
.TP 
\*(T<\fB\-l\fR\*(T>, \*(T<\fB\-\-led\fR\*(T> \fINUM\fR
Set LED status. Possible values for \fINUM\fR are:

\fBLED Status Codes\fR
.TS
r l.
T{
Num
T}	T{
Behavior
T}
.T&
r l.
T{
0
T}	T{
off
T}
T{
1
T}	T{
all blinking
T}
T{
2
T}	T{
1/top-left blink, then on
T}
T{
3
T}	T{
2/top-right blink, then on
T}
T{
4
T}	T{
3/bottom-left blink, then on
T}
T{
5
T}	T{
4/bottom-right blink, then on
T}
T{
6
T}	T{
1/top-left on
T}
T{
7
T}	T{
2/top-right on
T}
T{
8
T}	T{
3/bottom-left on
T}
T{
9
T}	T{

NEWS
## xboxdrv 0.8.14 - (2025-02-14)

* fixed evdev force feedback duration and profile
* removed guix, make, scons, etc

## xboxdrv 0.8.13 - (2024-10-09)

* fixed systemd path
* fixed uinput write error


## xboxdrv 0.8.12 - (2024-10-07)

* added meson build system
* fixed segfault


## xboxdrv 0.8.11 - (2024-04-06)

* removed boost dependency


## xboxdrv 0.8.10 - (2022-08-04)

Cherry-picks from chewi and cebtenzzre repos.
Excludes changes that broke input remapping.

* fixed controller disconnect crash/hang
* fixed daemon exit on controller disconnect and other issues
* fixed infinite force feedback
* fixed mismatched button identifiers
* fixed rumble-gain
* removed some broken URLs
* added GPD Win 2
* added Razer Sabertooth
* added Thrustmaster T-Wireless 3-in-1
* added "fake" wireless XBox360 receiver (045e:02a9)
* added hacky direct evdev force-feedback


## xboxdrv 0.8.9 - (2022-05-21)

Based on stable branch with startup delay fix.

* fixed startup delay
* fixed Chatpad memset bug
* improved error messages
* added actual option strings to ParsedOptions
* added alternative Afterglow AX.1 id


## xboxdrv 0.8.8 - (2015-11-09)

* use session dbus even for user root, except when DISPLAY is not set
* fixed issue in to_float() triggering assert()
* fixed vendor/product not getting set for virtual device in daemon mode
* fixed \r\n newlines causing errors in INIParser, ignore \r like whitespace
* added device support for Rock Candy Gamepad Wired Controller, Mad
  Catz Street Fighter IV SE Fighting Stick, MadCatz GamePad, Hori
  Fighting Edge, Elcom JC-U3613M, Mad Catz Fightpad, Mad Cats Ghost
  Recon FS GamePad, Mad Catz SFxT Fightstick Pro, Razer Sabertooth,
  FUS1ON Tournament Controller, PowerA Pro Ex and others


## xboxdrv 0.8.7 - (2015-10-28)

* fixed USB cleanup issue on exit


## xboxdrv 0.8.6 - (2015-10-25)

* Buttonevent exec: double fork to avoid zombies
* fixed axis handling in FourWayRestrictorModifier
* fixed device disconnect issues
* fixed segfault due to incorrect shutdown of USB subsystem
* fixed some compiler warnings/errors
* ignore Play'n Charge kit instead of error out
* added device support for Thrustmaster Gamepad GP XID, Razer
  Sabertooth, Gamestop BB-070, Afterglow AX.1, MLG Pro Circuit
  Controller (Xbox), Rock Candy, Saitek p3600, Hori GEM Xbox
  controller


## xboxdrv 0.8.5 - (2013-04-07)

* fixed errors on shutdown related to libusbx
* updated button mapping for --mimic-xpad-wireless and --mimic-xpad
* workaround for evdev values that don't fall into the given min/max
* added some support for Playstation 3 LED and rumble
* added support for Batarang Xbox 360 controller
* added support for Elite Glow
* added support for Hori Real Arcade Pro VX-SA
* added support for Hori SOULCALIBUR V Stick
* added support for Hori XBOX 360 EX 2 with Turbo
* added support for Logic3 Controller
* added support for Logitech F310
* added support for Logitech F510
* added support for Logitech F710
* added support for Razer Onza Tournament Edition
* added support for Street Fighter IV FightStick TE
* added support for Thrustmaster, Inc. GPX Controller
* added support for Xbox Airflo wired controller


## xboxdrv 0.8.4 - (2012-01-24)

* fixed readded lost man-page
* added documentation for --dbus and --ff-device


## xboxdrv 0.8.3 - (2012-01-23)

* added --dbus option to switch between system and session bus
* added sequence-key, same as cycle-key but doesn't wrap around
* added workaround for libudev older then v150
* added --ff-device DEV to select force-feedback device
* fixed force feedback getting on the wrong evdev
* fixed Xbox360 wireless LED not getting properly set on late plugin


## xboxdrv 0.8.2 - (2011-08-07)

* added click-press, click-release, click-both button filter
* added delay button filter
* added simple Python script for response curve generation
* added PDF Afterglow AX.1 Xbox360 controller support
* fixed swapped keys in KeyAxisEventHandler


## xboxdrv 0.8.1 - (2011-07-20)

* added const axis and button filter
* added 'generic-usb' controller type for debugging
* added cycle-key support to cycle through a sequence of buttons
* added proper merging of multiple output events to the same target
* added support for Street Fighter IV fightpad
* added support for Tron Xbox 360 controller
* added support for an unlimited number shift buttons
* fixed axis-shift getting ignored when another button is pressed
* fixed keys not getting pressed in --ui-axismap
* fixed rumble messages getting delayed
* reenabled --priority realtime


## xboxdrv 0.8.0 - (2011-05-26)

* added --no-dbus option
* added ABS/axis support to macros
* added Street Fighter Fightpad support
* added example config for Microsoft Ergonomic 4000 zoom wheel
* added example config for Star Wars: Tie Fighter
* added rel-repeat value of -1 for sending an event only once
* added support for chatpad on bcdVersion 1.14 controller
* added support for relative filenames for macros
* fixed daemon mode
* fixed deadzone filter issues with MIN:MAX:SMOOTH
* fixed endian issues in Xbox360 and Xbox360 wireless controller
* fixed wireless controller not getting reset to neutral position on disconnected
* changed the meaning of --timeout 0
* removed boost::thread dependency
* switched from threads to asynchronous USB handling and glib


## xboxdrv 0.7.3 - (2011-03-16)

* fixed installation of xboxdrvctl
* added special REL repeat value '-1' to avoid jaggy mouse emulation movement
* added "rel-repeat:" axis-event for better scroll wheel emulation
* added --no-dbus, to disable D-Bus support
* added xbmc.xboxdrv, a config fine tuned for XBMC
* added zsnes.xboxdrv, a config file for ZSNES
* reenabled USBReadThread to work around ignored input events
* '--daemon' is broken in this release


## xboxdrv 0.7.2 - (2011-03-06)

* added --mimic-xpad-wireless
* added --device-usbid and --device-usbids
* added D-Bus support to the daemon
* added xboxdrvctl to control the daemon

src/xboxdrv.cpp
/*
**  Xbox/Xbox360 USB Gamepad Userspace Driver
**  Copyright (C) 2008 Ingo Ruhnke <grumbel@gmail.com>
**
**  This program is free software: you can redistribute it and/or modify
**  it under the terms of the GNU General Public License as published by
**  the Free Software Foundation, either version 3 of the License, or
**  (at your option) any later version.
**
**  This program is distributed in the hope that it will be useful,
**  but WITHOUT ANY WARRANTY; without even the implied warranty of
**  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
**  GNU General Public License for more details.
**
**  You should have received a copy of the GNU General Public License
**  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "xboxdrv.hpp"

#include <sched.h>
#include <sys/types.h>
#include <unistd.h>

#include <cerrno>
#include <csignal>
#include <cstdio>
#include <cstring>
#include <format>
#include <iostream>
#include <stdexcept>
#include <string>

#include "command_line_options.hpp"
#include "controller_factory.hpp"
#include "controller_thread.hpp"
#include "dummy_message_processor.hpp"
#include "evdev_controller.hpp"
#include "evdev_helper.hpp"
#include "helper.hpp"
#include "raise_exception.hpp"
#include "uinput_message_processor.hpp"
#include "usb_gsource.hpp"
#include "usb_helper.hpp"
#include "usb_subsystem.hpp"
#include "word_wrap.hpp"
#include "xboxdrv_daemon.hpp"
#include "xboxdrv_main.hpp"

// Some ugly global variables, needed for sigint catching
bool global_exit_xboxdrv = false;

void Xboxdrv::run_list_controller() {
  int ret = libusb_init(NULL);
  if (ret != LIBUSB_SUCCESS) {
    raise_exception(std::runtime_error,
                    "libusb_init() failed: " << usb_strerror(ret));
  }

  libusb_device** list;
  ssize_t num_devices = libusb_get_device_list(NULL, &list);

  int id = 0;
  std::cout << " id | wid | idVendor | idProduct | Name" << std::endl;
  std::cout << "----+-----+----------+-----------+-----------------------------"
               "---------"
            << std::endl;

  for (ssize_t dev_it = 0; dev_it < num_devices; ++dev_it) {
    libusb_device* dev = list[dev_it];
    libusb_device_descriptor desc;

    // FIXME: we silently ignore failures
    if (libusb_get_device_descriptor(dev, &desc) == LIBUSB_SUCCESS) {
      for (int i = 0; i < xpad_devices_count; ++i) {
        if (desc.idVendor == xpad_devices[i].idVendor &&
            desc.idProduct == xpad_devices[i].idProduct) {
          if (xpad_devices[i].type == GAMEPAD_XBOX360_WIRELESS) {
            for (int wid = 0; wid < 4; ++wid) {
              std::cout << std::format(
                               " {:2d} |  {:2d} |   {:#04x} |    {:#04x} | "
                               "{:s} (Port: {:d})",
                               id, wid, int(xpad_devices[i].idVendor),
                               int(xpad_devices[i].idProduct),
                               xpad_devices[i].name, wid)
                        << std::endl;
            }
          } else {
            std::cout << std::format(
                             " {:2d} |  {:2d} |   {:#04x} |    {:#04x} | {:s}",
                             id, 0, int(xpad_devices[i].idVendor),
                             int(xpad_devices[i].idProduct),
                             xpad_devices[i].name)
                      << std::endl;
          }
          id += 1;
          break;
        }
      }
    }
  }

  if (id == 0) std::cout << "\nno controller detected" << std::endl;

  libusb_free_device_list(list, 1 /* unref_devices */);
}

void Xboxdrv::run_list_supported_devices() {
  for (int i = 0; i < xpad_devices_count; ++i) {
    std::cout << std::format("{:s} {:#04x} {:#04x} {:s}\n",
                             gamepadtype_to_string(xpad_devices[i].type),
                             int(xpad_devices[i].idVendor),
                             int(xpad_devices[i].idProduct),
                             xpad_devices[i].name);
  }
}

bool xpad_device_sorter(const XPadDevice& lhs, const XPadDevice& rhs) {
  if (lhs.idVendor < rhs.idVendor) {
    return true;
  } else if (lhs.idVendor == rhs.idVendor) {
    return lhs.idProduct < rhs.idProduct;
  } else {
    return false;
  }
}

void Xboxdrv::run_list_supported_devices_xpad() {
  std::vector<XPadDevice> sorted_devices(xpad_devices_count);
  std::copy_n(xpad_devices, xpad_devices_count, sorted_devices.begin());
  std::sort(sorted_devices.begin(), sorted_devices.end(), xpad_device_sorter);

  for (int i = 0; i < xpad_devices_count; ++i) {
    std::cout << std::format(
        "{{ {:#04x}, {:#04x}, \"{:s}\", {:s} }},\n",
        int(sorted_devices[i].idVendor), int(sorted_devices[i].idProduct),
        sorted_devices[i].name,
        gamepadtype_to_macro_string(sorted_devices[i].type));
  }
}

void Xboxdrv::run_help_devices() {
  std::cout << " idVendor | idProduct | Name" << std::endl;
  std::cout << "----------+-----------+---------------------------------"
            << std::endl;
  for (int i = 0; i < xpad_devices_count; ++i) {
    std::cout << std::format("   {:#04x} |    {:#04x} | {:s}",
                             int(xpad_devices[i].idVendor),
                             int(xpad_devices[i].idProduct),
                             xpad_devices[i].name)
              << std::endl;
  }
}

void Xboxdrv::print_copyright() const {
  WordWrap wrap(get_terminal_width());
  wrap.para("xboxdrv " PACKAGE_VERSION);
  wrap.para("Copyright © 2008-2011 Ingo Ruhnke <grumbel@gmail.com>");
  wrap.para(
      "Licensed under GNU GPL version 3 or later "
      "<http://gnu.org/licenses/gpl.html>");
  wrap.para("This program comes with ABSOLUTELY NO WARRANTY.");
  wrap.para(
      "This is free software, and you are welcome to redistribute it under "
      "certain "
      "conditions; see the file COPYING for details.");
  wrap.newline();
}

void Xboxdrv::run_main(const Options& opts) {
  if (!opts.quiet) {
    print_copyright();
  }

  USBSubsystem usb_subsystem;
  XboxdrvMain xboxdrv_main(opts);
  xboxdrv_main.run();
}

void Xboxdrv::run_daemon(Options& opts) {
  if (!opts.quiet) {
    print_copyright();
  }

  if (opts.usb_debug) {
#if LIBUSB_API_VERSION >= 0x01000106
    libusb_set_option(NULL, LIBUSB_OPTION_LOG_LEVEL, 3);
#else
    libusb_set_debug(NULL, 3);
#endif
  }

  if (opts.get_controller_slot().get_force_feedback()) {
    std::cerr << "Warning: Daemon mode does not support force-feedback, so it "
                 "has been disabled."
              << std::endl;
    opts.get_controller_slot().set_force_feedback(false);
  }

  if (!opts.detach) {
    USBSubsystem usb_subsystem;
    XboxdrvDaemon daemon(opts);
    daemon.run();
  } else {
    pid_t pid = fork();

    if (pid < 0) {  // fork error
      raise_exception(std::runtime_error,
                      "failed to fork(): " << strerror(errno));
    } else if (pid > 0) {  // parent, just exit
      _exit(EXIT_SUCCESS);
    } else {  // child, run daemon
      pid_t sid = setsid();

      if (sid == static_cast<pid_t>(-1)) {
        raise_exception(std::runtime_error,
                        "failed to setsid(): " << strerror(errno));
      } else {
        if (chdir("/") != 0) {
          raise_exception(std::runtime_error,
                          "failed to chdir(\"/\"): " << strerror(errno));
        } else {
          USBSubsystem usb_subsystem;
          XboxdrvDaemon daemon(opts);
          daemon.run();
        }
      }
    }
  }
}

void Xboxdrv::run_list_enums(uint32_t enums) {
  const int terminal_width = get_terminal_width();

  WordWrap wrap(terminal_width);

  if (enums & Options::LIST_ABS) {
    wrap.println("EV_ABS:");
    wrap.para("  ", string_join(evdev_abs_names.get_names(), ", "));
    wrap.newline();
  }

  if (enums & Options::LIST_REL) {
    wrap.println("EV_REL:");
    wrap.para("  ", string_join(evdev_rel_names.get_names(), ", "));
    wrap.newline();
  }

  if (enums & Options::LIST_KEY) {
    wrap.println("EV_KEY:");
    wrap.para("  ", string_join(evdev_key_names.get_names(), ", "));
    wrap.newline();
  }

  if (enums & Options::LIST_X11KEYSYM) {
    std::vector<std::string> lst;
    for (X11KeysymEnum::const_iterator i = get_x11keysym_names().begin();
         i != get_x11keysym_names().end(); ++i) {
      lst.push_back(i->second);
    }
    wrap.println("X11Keysym:");
    wrap.para("  ", string_join(lst, ", "));
    wrap.newline();
  }

  if (enums & Options::LIST_AXIS) {
    std::vector<std::string> lst;
    for (int i = 1; i < XBOX_AXIS_MAX; ++i) {
      lst.push_back(axis2string(static_cast<XboxAxis>(i)));
    }
    wrap.println("XboxAxis:");
    wrap.para("  ", string_join(lst, ", "));
    wrap.newline();
  }

  if (enums & Options::LIST_BUTTON) {
    std::vector<std::string> lst;
    for (int i = 1; i < XBOX_BTN_MAX; ++i) {
      lst.push_back(btn2string(static_cast<XboxButton>(i)));
    }
    wrap.println("XboxButton:");
    wrap.para("  ", string_join(lst, ", "));
    wrap.newline();
  }
}

Xboxdrv::Xboxdrv() {}

Xboxdrv::~Xboxdrv() {}

void Xboxdrv::set_scheduling(const Options& opts) {
  if (opts.priority == Options::kPriorityRealtime) {
    // try to set realtime priority when root, as user there doesn't
    // seem to be a way to increase the priority
    log_info("enabling realtime priority scheduling");

    int policy = SCHED_RR;

    struct sched_param param;
    memset(&param, 0, sizeof(struct sched_param));
    param.sched_priority = sched_get_priority_max(policy);

    // we don't try SCHED_OTHER for users as min and max priority is
    // 0 for that, thus we can't change anything with that

    int ret;
    if ((ret = sched_setscheduler(getpid(), policy, &param)) != 0) {
      raise_exception(std::runtime_error,
                      "sched_setschedparam() failed: " << strerror(errno));
    }
  }
}

int Xboxdrv::main(int argc, char** argv) {
  try {
    Options opts;

    CommandLineParser cmd_parser;
    cmd_parser.parse_args(argc, argv, &opts);

    set_scheduling(opts);

    switch (opts.mode) {
      case Options::PRINT_HELP_DEVICES:
        run_help_devices();
        break;

      case Options::RUN_LIST_SUPPORTED_DEVICES:
        run_list_supported_devices();
        break;

      case Options::RUN_LIST_SUPPORTED_DEVICES_XPAD:
        run_list_supported_devices_xpad();
        break;

      case Options::PRINT_VERSION:
        cmd_parser.print_version();
        break;

      case Options::PRINT_HELP:
        cmd_parser.print_help();
        break;

      case Options::PRINT_LED_HELP:
        cmd_parser.print_led_help();
        break;

      case Options::RUN_DEFAULT:
        run_main(opts);
        break;

      case Options::PRINT_ENUMS:
        run_list_enums(opts.list_enums);
        break;

      case Options::RUN_DAEMON:
        run_daemon(opts);
        break;

      case Options::RUN_LIST_CONTROLLER:
        run_list_controller();
        break;
    }
  } catch (const std::exception& err) {
    std::cout << "\n-- [ ERROR ] "
                 "------------------------------------------------------\n"
              << err.what() << std::endl;
  }

  return 0;
}

/* EOF */
